File: src/hbd/admin/routers/generations.py
"""``GET /api/generations`` and ``GET /api/generations/{attempt_id}`` — the render ledger.

**The wire model is :mod:`hbd.admin.schemas.orders`' and this module declares none of its
own.** ``AttemptWireView`` is already the projection of a ``generation_attempts`` row — the
order detail screen renders it — and a ``schemas/generations.py`` holding a second one would
be two views of one table. Two views of one table drift: the next privacy fix lands on the
copy whose screen somebody was looking at, and the other one keeps shipping the field that
was supposed to have been withdrawn. There is exactly one projection of this row in this
codebase, :func:`~hbd.admin.schemas.orders.to_attempt_view`, so the masking it applies is
applied everywhere the row is read.

**``isOrphaned`` selects attempts with no order, and it deliberately cannot say which kind.**
``order_id`` is nullable with ``ON DELETE SET NULL`` (``models/generation_attempt.py``): a
name preview is rendered *before* an order exists, and an attempt whose order was deleted
keeps its tuning signal, which is the entire reason the FK sets null instead of cascading —
the bake-off data has to outlive the orders it was collected from. Both reach the query as
``order_id IS NULL`` and nothing in the row tells them apart. Reporting a guess about which
one an operator is looking at would be worse than reporting the fact that they look
identical, so this filter states one predicate and the docstring states its two causes.

**``costUsd`` and ``latencyMs`` are ``null``, not ``0``, and ``isInstrumented`` says which
kind of row it is.** The columns default to ``0.0`` and ``0`` and nothing in ``src/`` writes
either, so every row in production today is uninstrumented. Rendering the defaults as
"$0.00 / 0 ms" on the one screen an operator uses to decide what to spend would be a
confident lie — and once Phase 5 starts writing real numbers it would be indistinguishable
from a genuinely free, instantaneous vendor call. ``attempts._telemetry`` reads a zero as
"never written"; this router simply does not undo that.

**The guard is ``require_permission`` at the router**, which resolves to ``check_role``:
there is no subject to weigh a step-up against on a list, and RECORDS_READ asks for none.
§12.2 gives RECORDS_READ as **M** to all four roles *including OWNER*, so there is no
unmasked cell to choose between and neither handler takes an ``Admin`` — taking the operator
in order to ignore them would imply a masking decision that does not exist. Plaintext for
``name_candidate_text`` and ``stt_transcript`` arrives through ``POST /reveal`` in Phase 2,
never from here.
"""

from __future__ import annotations

from datetime import datetime
from typing import Annotated, Final
from uuid import UUID

from fastapi import APIRouter, Depends, Query

from hbd.admin.deps import API_PREFIX, Db, require_permission
from hbd.admin.errors import AdminProblem, ProblemError, unwrap
from hbd.admin.schemas.orders import AttemptsPage, AttemptWireView, to_attempt_view
from hbd.admin.schemas.page import Paging, page_meta
from hbd.admin.security.permissions import Permission
from hbd.contracts import NameStrategy
from hbd.db.admin.attempts import AttemptFilters, count_attempts, get_attempt, list_attempts
from hbd.db.admin.sql import TimeWindow, time_window
from hbd.db.enums import GenerationKind
from hbd.db.models.generation_attempt import ERROR_CODE_LENGTH, PROVIDER_LENGTH
from hbd.errors import ErrorCode

__all__ = ["GENERATIONS_PATH", "ATTEMPT_PATH", "build_generations_router"]

GENERATIONS_PATH: Final[str] = f"{API_PREFIX}/generations"
ATTEMPT_PATH: Final[str] = f"{GENERATIONS_PATH}/{{attempt_id}}"


def _invalid(message: str) -> ProblemError:
    """422 in the pipeline taxonomy — the code the rest of the system already uses for this."""
    return ProblemError(AdminProblem(code=ErrorCode.INVALID_INPUT, message=message))


def _aware(name: str, value: datetime | None) -> datetime | None:
    """Refuse a naive instant (§6.1): the column is ``timestamptz`` and would raise anyway."""
    if value is not None and value.tzinfo is None:
        raise _invalid(f"{name} must carry a UTC offset, e.g. 2026-08-30T12:00:00Z")
    return value


def _window(since: datetime | None, until: datetime | None) -> TimeWindow | None:
    """``from`` and ``to`` are a pair — both, or neither.

    A one-sided window has to invent its missing bound and both candidates are a lie about
    what was asked for. The epoch is a bound nobody typed, and ``now`` is re-evaluated on
    every request — so a keyset walk down a ``?from=`` result would quietly widen its own
    filter between page one and page two, which is the drift the opaque cursor exists to
    prevent. Refusing the half-window costs a caller one extra parameter and costs nobody a
    page of rows that silently changed shape underneath them.
    """
    start = _aware("from", since)
    end = _aware("to", until)
    if start is None and end is None:
        return None
    if start is None or end is None:
        raise _invalid("from and to are a pair: send both bounds of the window, or neither")
    return unwrap(time_window(start, end))


def build_query(
    kind: Annotated[list[GenerationKind] | None, Query()] = None,
    provider: Annotated[str | None, Query(max_length=PROVIDER_LENGTH)] = None,
    is_success: Annotated[bool | None, Query(alias="isSuccess")] = None,
    error_code: Annotated[
        str | None, Query(alias="errorCode", max_length=ERROR_CODE_LENGTH)
    ] = None,
    strategy: Annotated[NameStrategy | None, Query()] = None,
    is_orphaned: Annotated[bool | None, Query(alias="isOrphaned")] = None,
    since: Annotated[datetime | None, Query(alias="from")] = None,
    until: Annotated[datetime | None, Query(alias="to")] = None,
) -> AttemptFilters:
    """§6.7's filter set, as a dependency so the handler stays a straight line.

    Repeated ``kind`` parameters are OR within the field and AND across fields (§6.1).
    ``kind`` and ``strategy`` are typed as enums, so an unknown value is a 422 from FastAPI
    rather than a filter that quietly matches nothing; ``provider`` and ``errorCode`` are
    free-form because their vocabularies are a vendor's and the pipeline's respectively, so
    they are bounded by the column widths they will be compared against instead.
    """
    return AttemptFilters(
        kinds=tuple(kind or ()),
        provider=provider,
        is_success=is_success,
        error_code=error_code,
        strategy=strategy,
        is_orphaned=is_orphaned,
        window=_window(since, until),
    )


Filters = Annotated[AttemptFilters, Depends(build_query)]


def _not_found(attempt_id: UUID) -> ProblemError:
    """404 through the pipeline taxonomy — ``STATUS_BY_ERROR_CODE`` maps ``NOT_FOUND``.

    The id is echoed into ``details`` because it is a ``UUID`` FastAPI already parsed, not
    caller-chosen text: a malformed one never reaches this function.
    """
    return ProblemError(
        AdminProblem(
            code=ErrorCode.NOT_FOUND,
            message="no generation attempt has that id",
            details={"attemptId": str(attempt_id)},
        )
    )


def build_generations_router() -> APIRouter:
    """The generation surface. One permission, declared once, on the router."""
    router = APIRouter(
        tags=["generations"],
        dependencies=[Depends(require_permission(Permission.RECORDS_READ))],
    )

    @router.get(GENERATIONS_PATH)
    async def list_generations(
        db: Db,
        filters: Filters,
        paging: Paging,
        with_total: Annotated[bool, Query(alias="withTotal")] = False,
    ) -> AttemptsPage:
        """One keyset page of the render ledger, newest first.

        The count is a second query and it runs only when asked for: ``bounded_total`` stops
        at ``TOTAL_COUNT_CAP``, so it is cheap, but it is not free and no screen needs it to
        render the first page.
        """
        page = await list_attempts(db, filters=filters, request=paging)
        total = await count_attempts(db, filters=filters) if with_total else None
        return AttemptsPage(
            items=[to_attempt_view(attempt) for attempt in page.items],
            meta=page_meta(page, total),
        )

    @router.get(ATTEMPT_PATH)
    async def get_generation(db: Db, attempt_id: UUID) -> AttemptWireView:
        """One attempt by id. An id that matches nothing is a 404, never an empty body."""
        attempt = await get_attempt(db, attempt_id)
        if attempt is None:
            raise _not_found(attempt_id)
        return to_attempt_view(attempt)

    return router

File: admin-dashboard/src/api/generations.ts
/**
 * `GET /api/generations` and `GET /api/generations/{attemptId}` — the render ledger.
 *
 * Transcribed from `.openpencil-export/users-generations-openapi.json` (produced by
 * `hbd/admin/routers/generations.py` over `hbd/admin/schemas/orders.py` — there is exactly
 * ONE projection of a `generation_attempts` row in that codebase, so the masking it applies
 * is applied everywhere the row is read).
 *
 * ## Four things this wire says that a table must not flatten
 *
 * **`costUsd` and `latencyMs` are `null`, not `0`, and `isInstrumented` says which kind of
 * row this is.** The columns default to `0.0`/`0` and nothing writes them yet, so every row
 * in production today is uninstrumented. "$0.00 / 0 ms" on the one screen an operator uses
 * to decide what to spend is a confident lie, and once real numbers land it would be
 * indistinguishable from a genuinely free, instantaneous vendor call.
 *
 * **`isOrphaned` states one predicate and cannot say which of its two causes applies.**
 * `order_id IS NULL` is both a name preview rendered before any order existed and an attempt
 * whose order was deleted — the FK sets null rather than cascading precisely so the tuning
 * data outlives the orders it came from. Nothing in the row tells them apart, so the UI must
 * not guess.
 *
 * **`nameCandidate` is MASKED and `sttTranscriptChars` is a LENGTH.** The name is the
 * orthography a vendor was asked to sing; the transcript is a near-verbatim copy of the whole
 * song, so only its size is on this wire. Plaintext for either comes from
 * `POST /api/reveal` (`generation_attempts.name_candidate_text`,
 * `generation_attempts.stt_transcript`) and from nowhere else — and both are PAGED reveals,
 * charged per page against the daily conversation ceiling as well as the record one.
 *
 * **Purged is not missing.** `identityPurgedAt` and `textPurgedAt` are the retention clocks
 * for those two columns. A purged value renders as "purged {date}" — never blank, and never
 * behind a reveal affordance that cannot succeed.
 *
 * ## Roles
 *
 * Both routes sit on `records.read`, which §12.2 gives to all four roles as **M** (masked) —
 * OWNER included. There is no unmasked cell here for anyone, so neither handler even takes
 * the operator. Nothing on this screen is role-gated; the reveal dialog it opens is.
 */

import { z } from "zod";

import { request, type ApiResult } from "./client";
import { GENERATIONS_PREFIX } from "./constants";
import {
  appendEach,
  appendPage,
  appendParam,
  pageMetaSchema,
  queryOf,
  type PageRequest,
} from "./pagination";
import { languageSchema } from "./users";

/* `Language` is declared beside `UserView`, its first consumer, and imported here rather
   than declared twice: two spellings of one closed vocabulary drift, and the drift always
   shows up as a filter that quietly matches nothing. */
export { LANGUAGE_VALUES, languageSchema, type Language } from "./users";

/* -------------------------------------------------------------------------- */
/* Scalars                                                                     */
/* -------------------------------------------------------------------------- */

/** RFC 3339, `Z`-suffixed. A string, for `auth.ts`'s reason. */
const timestampSchema = z.string();

/* -------------------------------------------------------------------------- */
/* Enums — verbatim from the spec; a new member must be added here first        */
/* -------------------------------------------------------------------------- */

/**
 * `GenerationKind` — what the attempt was trying to make. `name_preview` and
 * `name_verification` are the two that legitimately have no order, which is what makes
 * `isOrphaned` a normal condition rather than a fault.
 */
export const GENERATION_KIND_VALUES = [
  "song",
  "song_inpaint",
  "greeting",
  "lyrics",
  "name_preview",
  "name_verification",
  "cover",
] as const;
export const generationKindSchema = z.enum(GENERATION_KIND_VALUES);
export type GenerationKind = z.infer<typeof generationKindSchema>;

/**
 * `NameStrategy` — which spelling of the recipient's name this attempt tried. The bake-off
 * axis: `nameCandidateRank`, `isNameVerified` and `matchConfidence` are only readable beside
 * it.
 */
export const NAME_STRATEGY_VALUES = [
  "canonical",
  "stripped",
  "ascii",
  "cyrillic",
  "hyphenated",
  "phonetic",
] as const;
export const nameStrategySchema = z.enum(NAME_STRATEGY_VALUES);
export type NameStrategy = z.infer<typeof nameStrategySchema>;

/**
 * The ADAPTER names this deployment writes — finer than the `Vendor` an invoice is issued
 * under (two ElevenLabs adapters, one bill). `openai-compat` is hyphenated and the others are
 * not; that is the literal in `providers/llm/openai_compat.py` and it is not a typo to tidy.
 *
 * **The tuple below is the whole vocabulary**, taken from `providers/**` and from nowhere
 * else. A vendor remembered from a plan document or from another product is not in this
 * system; a filter offering one would return an empty page for ever and read as an outage.
 *
 * The `fake_*` four are `HBD_USE_FAKE_PROVIDERS` runs: recorded rather than dropped, because
 * no rows at all cannot be told from an uninstrumented deploy.
 */
export const PROVIDER_VALUES = [
  "elevenlabs_music",
  "elevenlabs_tts",
  "elevenlabs_scribe",
  "openai-compat",
  "gemini",
  "fake_music",
  "fake_tts",
  "fake_llm",
  "fake_stt",
] as const;
/**
 * For the FILTER only — a picker offering a value the server has never seen returns an empty
 * page, so the vocabulary is closed where an operator chooses from it.
 *
 * `AttemptWireView.provider` is deliberately NOT parsed with this. The column is free-form
 * `varchar(64)` (the vocabulary is a vendor's, not ours), so a new adapter shipping tomorrow
 * must render as itself rather than turn every row on the screen into a `SCHEMA_DRIFT`
 * banner. Widen the tuple when an adapter is added; the wire keeps working meanwhile.
 */
export const providerSchema = z.enum(PROVIDER_VALUES);
export type Provider = z.infer<typeof providerSchema>;

/* -------------------------------------------------------------------------- */
/* GET /api/generations                                                        */
/* -------------------------------------------------------------------------- */

/** One row of the render ledger. */
export const attemptWireViewSchema = z.object({
  id: z.string().uuid(),
  /** `null` exactly when `isOrphaned` — see the header for the two causes. */
  orderId: z.string().uuid().nullable(),
  kind: generationKindSchema,
  /** Which stage instance within the order, and which try within that stage. */
  sequence: z.number().int(),
  attempt: z.number().int(),
  /** The adapter that made the call. Free-form on the wire; see `providerSchema`. */
  provider: z.string().nullable(),
  /** The vendor's own id for the job, for chasing it in their console. */
  providerRemoteId: z.string().nullable(),
  language: languageSchema.nullable(),
  isSuccess: z.boolean(),
  isOrphaned: z.boolean(),

  nameCandidateStrategy: nameStrategySchema.nullable(),
  nameCandidateRank: z.number().int().nullable(),
  /** `null` = never verified, which is not the same as verified-and-failed (`false`). */
  isNameVerified: z.boolean().nullable(),
  matchConfidence: z.number().nullable(),

  /** MASKED, like any other name. Plaintext only via a reveal, and it is a paged one. */
  nameCandidate: z.string().nullable(),
  /** Retention clock for `nameCandidate`. Non-null = "purged {date}", not "missing". */
  identityPurgedAt: timestampSchema.nullable(),
  /** LENGTH only — the transcript is a near-verbatim copy of the whole song. */
  sttTranscriptChars: z.number().int().nullable(),
  /** Retention clock for the transcript and the candidate text. */
  textPurgedAt: timestampSchema.nullable(),

  errorCode: z.string().nullable(),
  /** Our own operator prose plus a closed-vocabulary code. Never a customer's words. */
  errorMessage: z.string().nullable(),
  /** `null` = the code is unknown to the taxonomy, NOT "not retryable". */
  isRetryable: z.boolean().nullable(),

  /** USD, decimal. `null` = never instrumented; never render it as `$0.00`. */
  costUsd: z.number().nullable(),
  costSource: z.string().nullable(),
  /** Whole milliseconds. `null` = never instrumented. */
  latencyMs: z.number().int().nullable(),
  /** Whether this row's telemetry was ever written. False today for every row. */
  isInstrumented: z.boolean(),
  createdAt: timestampSchema,
});
export type AttemptWireView = z.infer<typeof attemptWireViewSchema>;

export const attemptsPageSchema = z.object({
  items: z.array(attemptWireViewSchema),
  meta: pageMetaSchema,
});
export type AttemptsPage = z.infer<typeof attemptsPageSchema>;

/* -------------------------------------------------------------------------- */
/* Filters                                                                     */
/* -------------------------------------------------------------------------- */

/**
 * §6.7's filter set. Every field is optional; an omitted one is not sent, and each boolean is
 * a TRI-STATE — absent, `true` and `false` are three different questions.
 */
export interface GenerationsFilters {
  /** The bounded count as well. A second query; nothing needs it to render the first page. */
  readonly withTotal?: boolean;
  /** REPEATED on the wire: OR within the field, AND across fields. Empty = do not filter. */
  readonly kind?: readonly GenerationKind[];
  /**
   * Exact adapter name, one at a time — this one does NOT repeat. A free string because the
   * vocabulary is a vendor's; `PROVIDER_VALUES` is what a picker offers, not what the
   * parameter accepts. Cap the INPUT at `MAX_PROVIDER_CHARS`: an over-long value is a 422
   * naming the parameter, and truncating it here would hide that behind a page nobody asked
   * for.
   */
  readonly provider?: string | null;
  readonly isSuccess?: boolean | null;
  /**
   * Exact pipeline error code, one at a time. Free-form: the vocabulary is the pipeline's.
   * Cap the input at `MAX_ERROR_CODE_CHARS`, for `provider`'s reason.
   */
  readonly errorCode?: string | null;
  /** One strategy. An unknown value is a 422, not a filter that quietly matches nothing. */
  readonly strategy?: NameStrategy | null;
  /** `true` = attempts with no order at all. It cannot say WHICH kind of orphan. */
  readonly isOrphaned?: boolean | null;
  /** `created_at`, half-open `[from, to)`, RFC 3339. Either end may stand alone. */
  readonly from?: string | null;
  readonly to?: string | null;
}

function generationsQuery(filters: GenerationsFilters, page: PageRequest): string {
  const params = new URLSearchParams();
  if (filters.withTotal === true) params.append("withTotal", "true");
  appendEach(params, "kind", filters.kind);
  appendParam(params, "provider", filters.provider);
  appendParam(params, "isSuccess", filters.isSuccess);
  appendParam(params, "errorCode", filters.errorCode);
  appendParam(params, "strategy", filters.strategy);
  appendParam(params, "isOrphaned", filters.isOrphaned);
  appendParam(params, "from", filters.from);
  appendParam(params, "to", filters.to);
  appendPage(params, page);
  return queryOf(params);
}

/* -------------------------------------------------------------------------- */
/* The routes                                                                  */
/* -------------------------------------------------------------------------- */

/** The route templates, as a failure names them. */
export const GENERATIONS_ENDPOINT = {
  list: "GET /api/generations",
  detail: "GET /api/generations/{attemptId}",
} as const;

/** One keyset page of the render ledger, newest first. */
export function listGenerations(
  filters: GenerationsFilters,
  page: PageRequest,
  signal?: AbortSignal,
): Promise<ApiResult<AttemptsPage>> {
  return request({
    endpoint: GENERATIONS_ENDPOINT.list,
    path: `${GENERATIONS_PREFIX}${generationsQuery(filters, page)}`,
    schema: attemptsPageSchema,
    ...(signal === undefined ? {} : { signal }),
  });
}

/**
 * One attempt by id. An id that matches nothing is a 404, never an empty body — so a detail
 * panel opened on a stale link says "no such attempt" rather than drawing an empty shell.
 */
export function getGeneration(
  attemptId: string,
  signal?: AbortSignal,
): Promise<ApiResult<AttemptWireView>> {
  return request({
    endpoint: GENERATIONS_ENDPOINT.detail,
    path: `${GENERATIONS_PREFIX}/${encodeURIComponent(attemptId)}`,
    schema: attemptWireViewSchema,
    ...(signal === undefined ? {} : { signal }),
  });
}
